build_mt.py

```diff
@@ -1,599 +1,827 @@
-    
-    # percentage of objects that are part of this or previous snapshot
-    print("percentage of halos in light cone with and without progenitor info = ", \
-          np.sum(mask_lc_this_info)/len(mask_lc_this_info)*100., np.sum(mask_lc_this_noinfo)/len(mask_lc_this_noinfo)*100.)
-
-    # select halos with mt info that have had a light cone crossing
-    pos_this_info_lc = pos_this_info[mask_lc_this_info]
-    com_dist_this_info_lc = com_dist_this_info[mask_lc_this_info]
-    pos_prev_main_this_info_lc = pos_prev_main_this_info[mask_lc_this_info]
-    com_dist_prev_main_this_info_lc = com_dist_prev_main_this_info[mask_lc_this_info]
-    halo_ind_prev_main_this_info_lc = halo_ind_prev_main_this_info[mask_lc_this_info]
-    halo_ind_this_info_lc = halo_ind_this_info[mask_lc_this_info]
-    eligibility_this_info_lc = eligibility_this_info[mask_lc_this_info]
-    if 'Progenitors' in fields_mt:
-        start_progs_this_info_lc = start_progs_this_info[mask_lc_this_info]
-        num_progs_this_info_lc = num_progs_this_info[mask_lc_this_info]
+    # assert np.sum((chi_star > chi1) | (chi_star < chi2)) == 0, "Solution is out of bounds"
 
+    return chi_star, pos_star, vel_star, bool_star
 
-    
-    # START TESTING
-    '''
-    x_min = 0.
-    x_max = x_min+10.
-    
-    x = pos_this_info_lc[:,0]
-    choice = (x > x_min) & (x < x_max)
-    
-    y = pos_this_info_lc[choice,1]
-    z = pos_this_info_lc[choice,2]
 
-    plt.figure(1)
-    plt.scatter(y,z,color='dodgerblue',s=0.1,label='current objects')
+def get_one_header(merger_dir):
+    '''Get an example header by looking at one association file in a merger directory'''
+    fn = list(merger_dir.glob('associations*.asdf'))[0]
+    with asdf.open(fn) as af:
+        header = af['header']
+    return header
+
+
+def main(sim_name, z_start, z_stop, resume=False, plot=False):
+    # speed of light
+    c = 299792.458  # km/s
+    
+    merger_parent = Path("/mnt/gosling2/bigsims/merger")
+    merger_dir = merger_parent / sim_name
+    header = get_one_header(merger_dir)
+    
+    # simulation parameters
+    Lbox = header['BoxSize']
+    # location of the LC origins in Mpc/h
+    origins = np.array(header['LightConeOrigins']).reshape(-1,3)
+    origin = origins[0]
+
+    # directory where we save the final outputs
+    cat_lc_dir = Path("/mnt/gosling1/boryanah/light_cone_catalog/") / sim_name / "halos_light_cones/"
+    os.makedirs(cat_lc_dir, exist_ok=True)
+
+    # directory where we save the current state if we want to resume
+    os.makedirs(cat_lc_dir / "tmp", exist_ok=True)
+
+    # all redshifts, steps and comoving distances of light cones files; high z to low z
+    # LHG: these have to be recomputed per-sim. How is chi being determined? etaK from header?
+    zs_all = np.load("data_headers/redshifts.npy")
+    chis_all = np.load("data_headers/coord_dist.npy")
+    zs_all[-1] = float("%.1f" % zs_all[-1])  # LHG: I guess this is trying to match up to some filename or something?
```

Comment:
It's fine to buffer the redshifts of all the lightcone files if reading 800 headers is slow.  If it's only done once, not worth buffering, but I guess with resumes it might be more than once.

But note that the redshifts and chis of the light cone files changes from sim to sim.  And are the chis using etaK from the header?  I think that's what we'd prefer.

Reply:
the line that says 
zs_all[-1] = float("%.1f" % zs_all[-1])
is just because of the interpolation -- in one case we have 0.100001 and in the other 0.1 and I get an out of bounds error

Reply:
yep, yep, noted that chis and zs change -- I should perhaps read them explicitly!

this is using the Coordinate Distance Mpc/h which is what I thought we landed on with Daniel!

Reply:
That's fine about the interpolation, should just leave a comment since it's not obvious.

For the distance, I was just wondering what cosmology module was being used to compute the values.  I think we'd like to use Abacus's, which means reading etaK from the header and then getting the z=0 value from Abacus.Cosmology so you can do this calculation: https://github.com/abacusorg/abacus/blob/development/singlestep/Output/LightCones.cpp#L51

Reply:
Looking back at my notes, I see that the header is already populated with the CoordinateDistanceHMpc field.  Is that what you're using (just checking it's not coming from CAMB or astropy or something).